### PR TITLE
uNode board support - Incomplete

### PR DESCRIPTION
**Pull request by crashmatt**
includes makefiles, board drivers and nuttx configs for unode.
Significant changes to rcS to support startup for mtd flash.
Significant changes to mtd driver to support W25 memory formatted into unequal sizes and formatted with either character blocks or fat.

Significant risk with rcS script.  Moderate risk on mtd driver.

### CODE REVIEW

**Comment by kd0aij** [comment]
Matt, this looks pretty straightforward to me. Can I offer any help testing on Pixhawk or X2?

**Comment by crashmatt** [comment]
That would be great.  Key tests are if Pixhawk starts as expected with and without the microsd. Checks are:
Memories should mount in the right place.  
Parameters should be loaded from the right location (ramtron vs. flash).  
Logfile set to correct location
waypoints set to correct location 

I need to fix the Travis errors and the error report is not making any sense yet.

**Comment by LorenzMeier** [comment]
Looks in general good - I would advise removing the commented out code pieces (left-overs from porting) once done.

The Travis log is not hard to read. Just download the text log file and search for "error" - it will quickly take you to this GCC error:

```
/home/travis/build/PX4/Firmware/src/systemcmds/mtd/mtd.c:127:48: error: 'partition_names_default' undeclared here (not in a function)
 static const int n_partitions_default = sizeof(partition_names_default) / sizeof(partition_names_default[0]);
                                                ^
compilation terminated due to -Wfatal-errors.
```

Which looks like the compilation for a few other board configs broke as you did your MTD changes - running the build with just 'make' should turn up the same error for you. If not, try make clean && make.

**Comment by crashmatt** [comment]
Lorenz, I missed the px4 v1 case using eeprom. I have checked in a fix together with fixing the other mtd.c compile warnings.

Thanks for the hint on doing make for all.  I now found this to output everything into a logfile:
make 2>&1 | tee filetokeepitin.txt

**Comment by crashmatt** [comment]
I call this ready to go.
The risk on mtd has been minimized.  The defines make sure that the code is the same as it was before.
The change to rcS is only significant if the sd card is not found.  It then tries to mount mtd and tests for /fs/microsd.  It then sets the logfile location depending on it finding /fs/microsd on the mtd device.

File: src/drivers/boards/unode/unode_init.c [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+#endif
+
+__EXPORT int nsh_archinitialize(void)
+{
+
+	/* configure ADC pins */
+	stm32_configgpio(GPIO_ADC1_IN2);	/* BATT_VOLTAGE_SENS */
+	stm32_configgpio(GPIO_ADC1_IN3);	/* BATT_CURRENT_SENS */
+	stm32_configgpio(GPIO_ADC1_IN4);	/* VDD_5V_SENS */
+	// stm32_configgpio(GPIO_ADC1_IN10);	/* used by VBUS valid */
```
Comment (LorenzMeier) [inline src/drivers/boards/unode/unode_init.c]: Commented out code - I presume these pins don't exist?

File: src/drivers/boards/unode/unode_init.c [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+	// stm32_configgpio(GPIO_ADC1_IN11);	/* unused */
+	// stm32_configgpio(GPIO_ADC1_IN12);	/* used by MPU6000 CS */
+	stm32_configgpio(GPIO_ADC1_IN13);	/* FMU_AUX_ADC_1 */
+	stm32_configgpio(GPIO_ADC1_IN14);	/* FMU_AUX_ADC_2 */
+	stm32_configgpio(GPIO_ADC1_IN15);	/* PRESSURE_SENS */
+
+	/* configure power supply control/sense pins */
+	stm32_configgpio(GPIO_VDD_5V_PERIPH_EN);
+	stm32_configgpio(GPIO_VDD_3V3_SENSORS_EN);
+//	stm32_configgpio(GPIO_VDD_BRICK_VALID);
```
Comment (LorenzMeier) [inline src/drivers/boards/unode/unode_init.c]: Commented out code.

File: src/drivers/boards/unode/unode_spi.c [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+__EXPORT void stm32_spi1select(FAR struct spi_dev_s *dev, enum spi_dev_e devid, bool selected)
+{
+	/* SPI select is active low, so write !selected to select the device */
+
+	switch (devid) {
+	case PX4_SPIDEV_GYRO:
+		/* Making sure the other peripherals are not selected */
+		stm32_gpiowrite(GPIO_SPI_CS_GYRO, !selected);
+//		stm32_gpiowrite(GPIO_SPI_CS_ACCEL_MAG, 1);
```
Comment (LorenzMeier) [inline src/drivers/boards/unode/unode_spi.c]: Commented out code.

File: src/drivers/boards/unode/unode_spi.c [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+__EXPORT void stm32_spi4select(FAR struct spi_dev_s *dev, enum spi_dev_e devid, bool selected)
+{
+	/* SPI select is active low, so write !selected to select the device */
+
+	switch (devid) {
+	case PX4_SPIDEV_EXT0:
+		/* Making sure the other peripherals are not selected */
+		stm32_gpiowrite(GPIO_SPI_CS_EXT0, !selected);
+		stm32_gpiowrite(GPIO_SPI_CS_EXT1, 1);
+//		stm32_gpiowrite(GPIO_SPI_CS_EXT2, 1);
```
Comment (LorenzMeier) [inline src/drivers/boards/unode/unode_spi.c]: commented out code.

File: src/drivers/boards/unode/unode_usb.c [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file px4fmu_usb.c
```
Comment (LorenzMeier) [inline src/drivers/boards/unode/unode_usb.c]: comment lies

File: src/drivers/drv_gpio.h [hunk 6]
```diff
@@ -114,9 +114,35 @@
 /* no GPIO driver on the PX4_STM32F4DISCOVERY board */
 #endif
 
+
+#ifdef CONFIG_ARCH_BOARD_UNODE
+/*
+ * PX4FMUv2 GPIO numbers.
```
Comment (LorenzMeier) [inline src/drivers/drv_gpio.h]: comment lies.

File: src/drivers/drv_gpio.h [hunk 7]
```diff
@@ -114,9 +114,35 @@
...
+ * There are no alternate functions on this board.
+ */
+# define GPIO_SERVO_1		(1<<0)		/**< servo 1 output */
+# define GPIO_SERVO_2		(1<<1)		/**< servo 2 output */
+# define GPIO_SERVO_3		(1<<2)		/**< servo 3 output */
+# define GPIO_SERVO_4		(1<<3)		/**< servo 4 output */
+# define GPIO_SERVO_5		(1<<4)		/**< servo 5 output */
+# define GPIO_SERVO_6		(1<<5)		/**< servo 6 output */
+
+# define GPIO_5V_PERIPH_EN	(1<<6)		/**< PA8 - !VDD_5V_PERIPH_EN */
```
Comment (LorenzMeier) [inline src/drivers/drv_gpio.h]: These pins exist, yes?

**Comment by crashmatt** [comment]
Lorenz,
Do I need to tidy everything before it is accepted into upstream?

It is a significant chunk of work to test adc, spi gpio etc. With more changes to upstream I risk having to re-write everything for the third time.  For now there are some rough changes just to get the board running at all.  I have commented many lines instead of removing them so I can clearly see what I have done.

What do you recommend?

/Matt

**Comment by LorenzMeier** [comment]
Generally getting past the gate to master means you have something that is "done". That is the same policy we apply - in fact anything that makes it onto master these days is bench and flight tested.

If you want to get something into master that is fundamentally work in progress there are a few guidelines:

  * Contributed, active code should be tested. This is to save both you and me some work: You because you don't have to come back to it, me because we don't have untested code in the repository
  * If there are WIP items they should be clearly commented - so for the lines of code you commented out, each of them (or each group) should come with a comment saying why its there and on what condition it can later be removed.

If you look at the px4fmu-v2_test config you will find the tests tool (src/systemcmds/tests) which includes test routines to get ADC, sensors and PWM out tested on the bench. I would appreciate if you could provide the test output for those here, and with the comments added / fixed I think it would be in a good shape to be in master.

If that's beyond reach the code should at least clearly be commented to make sure anyone trying to base off your work knows what is tested and what is not. Because once its merged to master, the whole context you know about (and this thread) are gone.

**Comment by crashmatt** [comment]
Lorenz,  thanks for your review.  I agree with how you are controlling the upstream quality. 

It will take a while to fix everything.  Pull request closed so it does not need to be maintained.

**Pull request closed without merging**